iroh-bytes/src/provider.rs

```diff
@@ -211,6 +213,133 @@ pub trait CustomGetHandler: Send + Sync + Debug + 'static {
     ) -> BoxFuture<'static, anyhow::Result<GetRequest>>;
 }
 
+/// A custom collection parser that allows the user to define what a collection is.
+///
+/// A collection can be anything that contains an ordered sequence of blake3 hashes.
+/// Some collections store links with a fixed size and therefore allow efficient
+/// skipping. Others store links with a variable size and therefore only allow
+/// sequential access.
+///
+/// This API tries to accomodate both use cases. For collections that do not allow
+/// efficient random access, the [`LinkStream::skip`] method can be implemented by just repeatedly
+/// calling `next`.
+///
+/// For collections that do allow efficient random access, the [`LinkStream::skip`] method can be
+/// used to move some internal offset.
+pub trait CollectionParser: Send + Debug + Clone + 'static {
+    /// Parse a collection with this parser
+    fn parse<'a, R: AsyncSliceReader + 'a>(
+        &'a self,
+        format: u64,
+        reader: R,
+    ) -> LocalBoxFuture<'a, anyhow::Result<(Box<dyn LinkStream>, CollectionStats)>>;
```

Comment:
Does this mean that at parse time you need to know the `CollectionStats`?  What if you have to iterate through the entire collection to know how many links are in it?  What if you don't have the metadata to figure out the total blob size?